### PR TITLE
Add fit status fields to output container of MuonIntensityFitter

### PR DESCRIPTION
**Pull request by jstvdk**
Added field for the storing information about muon intensity fit convergence.
Related issue #2359

### CODE REVIEW

File: ctapipe/containers.py [hunk 1]
```diff
@@ -984,6 +984,7 @@ class MuonEfficiencyContainer(Container):
     impact_x = Field(nan * u.m, "impact parameter x position")
     impact_y = Field(nan * u.m, "impact parameter y position")
     optical_efficiency = Field(nan, "optical efficiency muon")
+    fit_convergence = Field(nan, "convergence of the fit")
```
Comment (maxnoe) [inline ctapipe/containers.py:987]: I'd calls this, in line with other such definitions, `is_valid`.

We could also add other fields with more specific information, e.g. using these fields:

```
minuit.fmin.has_parameters_at_limit
minuit.fmin.has_valid_parameters
minuit.fmin.is_valid
```

and the likelihood value is available as:

```
minuit.fval
```

While we are at it, we could also store the uncertainties:

```
minuit.errors
```

Comment (maxnoe) [reply ctapipe/containers.py:987]: Checking the minuit docs, `has_valid_parameters` and `is_valid` are aliases.

So we could just stick with `is_valid` and `parameters_at_limit`.

Comment (maxnoe) [reply ctapipe/containers.py:987]: Default should also be `False`, not `nan`

[the remark this answers is not in the document]
Comment (jstvdk) [reply ctapipe/containers.py]: Thanks, implemented

Comment (jstvdk) [reply ctapipe/containers.py]: Thanks, implemented

Comment (maxnoe) [reply ctapipe/containers.py]: You don't yet fill the `likelihood_value` and the `parameters_at_limit` fields

File: ctapipe/image/muon/intensity_fitter.py [hunk 2]
```diff
@@ -536,6 +536,9 @@ def __call__(self, tel_id, center_x, center_y, radius, image, pedestal, mask=Non
         # Perform minimisation
         minuit.migrad()
 
+        # Check for convergence
+        validation = minuit.valid
```
Comment (maxnoe) [inline ctapipe/image/muon/intensity_fitter.py:540]: This is very nitpicky, but there is no reason to create the local variable here, just do

```
is_valid=minuit.valid
```

below where the container is created.

File: ctapipe/image/muon/intensity_fitter.py [hunk 3]
```diff
@@ -545,4 +548,5 @@ def __call__(self, tel_id, center_x, center_y, radius, image, pedestal, mask=Non
             impact_y=result["impact_parameter"] * np.sin(result["phi"]) * u.m,
             width=u.Quantity(np.rad2deg(result["ring_width"]), u.deg),
             optical_efficiency=result["optical_efficiency_muon"],
+            fit_convergence=validation
```
Comment (maxnoe) [inline ctapipe/image/muon/intensity_fitter.py:551]: Still the old name, will fail now

**Comment by maxnoe** [comment]
Could you add a test for the new fields?

This could be simply checking

```
assert result.is_valid
assert not result.parameters_at_limit
assert np.isfinite(result.likelihood_value)
```

in one of the tests for now

**Review by maxnoe — APPROVED** [review]

**Comment by maxnoe** [comment]
The only thing missing now is a changelog entry.

You can add one by creating a file `docs/changes/2381.feature.rst` (the number of this PR + the category of the change) adding a small description mentioning the new fields.

**Review by maxnoe — COMMENTED** [review]
Looks good! Thank you very much.

The pre-commit check is complaining about some style issues.

To prevent that in the future, you can locally run `pre-commit install`

After that, you can now run to fix it for this PR:

```
pre-commit run --files $(git diff origin/main --name-only)
```
```

**Comment by jstvdk** [comment]
> To prevent that in the future, you can locally run `pre-commit install`
> 
> After that, you can now run to fix it for this PR:
> 
> ```
> pre-commit run --files $(git diff origin/main --name-only)
> ```

Thank you for the explanation

File: docs/changes/2381.feature.rst [hunk 4]
```diff
@@ -0,0 +1 @@
+Added new fields to the `MuonEfficiencyContainer` - `is_valid` to check if fit converged successfully, `parameters_at_limit` to check if parameters were fitted close to a bound and `likelihood_value` which represents cost function value atthe minimum. These fields were added to the output of the `MuonIntensityFitter`
```
Comment (maxnoe) [inline docs/changes/2381.feature.rst:1]:
```suggestion
Added new fields to the ``MuonEfficiencyContainer`` - ``is_valid`` to check if fit converged successfully, ``parameters_at_limit`` to check if parameters were fitted close to a bound and ``likelihood_value`` which represents cost function value at the minimum. These fields were added to the output of the ``MuonIntensityFitter``
```

In restructured text, you need to use two back-ticks for code, a single back-tick is interpreted as a reference and the docs builds now complains about undefined references.

So you need to repace all single ticks here with double ticks.

**Comment by maxnoe** [comment]
@jstvdk The correction for the style issues is still missing, that's all I guess.

Need any help with fixing those?

**Comment by jstvdk** [comment]
> @jstvdk The correction for the style issues is still missing, that's all I guess.
> 
> Need any help with fixing those?

Yes please, I didn't understood this completely.

**Comment by maxnoe** [comment]
> Yes please, I didn't understand this completely.

The CI runs the same checks you can run locally. What is currently failing is the check for the style guide, which can be solved by auto-formatting the code using the pre-commit config and committing the result.

Here you can see the error:
https://github.com/cta-observatory/ctapipe/actions/runs/5531696637/jobs/10093041566

To fix them, you can run:

`pre-commit run --files $(git diff origin/main --name-only)`

And then add / commit the changes and push

**Comment by jstvdk** [comment]
> To fix them, you can run:
> 
> `pre-commit run --files $(git diff origin/main --name-only)`
> 
> And then add / commit the changes and push

Thanks, I tried this solution, but it is not changing and pushing, because "Everything is up-to-date".

**Comment by maxnoe** [comment]
Ah, sorry, I forgot that you are working on a fork.

You need to compare to the main branch in this repository. Assuming you have it added as `upstream`, e.g. via:

```
git remote add upstream https://github.com/cta-observatory/ctapipe
```

To update and check:
```
git fetch upstream
git diff upstream/main --name-only
```

Then run with upstream instead of origin:
```
pre-commit run --files $(git diff upstream/main --name-only)
```

**Comment by jstvdk** [comment]
> Then run with upstream instead of origin:

Thank you, I made it.

**Review by maxnoe — APPROVED** [review]

File: ctapipe/image/muon/tests/test_intensity_fit.py [hunk 5]
```diff
@@ -89,6 +89,9 @@ def test_muon_efficiency_fit(prod5_lst):
     assert u.isclose(result.impact, impact_parameter, rtol=0.05)
     assert u.isclose(result.width, ring_width, rtol=0.05)
     assert u.isclose(result.optical_efficiency, efficiency, rtol=0.05)
+    assert result.is_valid
+    assert not result.parameters_at_limit
+    assert np.isfinite(result.likelihood_value)
```
Comment (Tobychev) [inline ctapipe/image/muon/tests/test_intensity_fit.py:94]: Is is very hard to change this test so it checks that the likelihood is reasonably close to a known value?

Comment (maxnoe) [reply ctapipe/image/muon/tests/test_intensity_fit.py:94]: Testing a particular value will be hard, yes. Maybe testing that the value is lower for a fit with the correct ring parameters passed in than for one where we don't have ring center / radius values?

**Review by kosack — APPROVED** [review]

**Pull request merged**